WordPress/Classes/Services/SiteAssemblyService.swift

```diff
@@ -4,26 +4,178 @@ import Foundation
+
+    /// The site creation request that's been enqueued.
+    private var creationRequest: SiteCreationRequest?
+
+    /// This handler is called with changes to the site assembly status.
+    private var statusChangeHandler: SiteAssemblyStatusChangedHandler?
+
+    /// The most recently created blog corresponding to the site creation request; `nil` otherwise.
+    private(set) var createdBlog: Blog?
+
+    // MARK: LocalCoreDataService
+
+    override init(managedObjectContext context: NSManagedObjectContext) {
+        self.accountService = AccountService(managedObjectContext: context)
+        self.blogService = BlogService(managedObjectContext: context)
+
+        let api: WordPressComRestApi
+        if let wpcomApi = accountService.defaultWordPressComAccount()?.wordPressComRestApi {
+            api = wpcomApi
+        } else {
+            api = WordPressComRestApi(userAgent: WPUserAgent.wordPress())
+        }
+        self.remoteService = WordPressComServiceRemote(wordPressComRestApi: api)
+
+        super.init(managedObjectContext: context)
+    }
+
+    // MARK: SiteAssemblyService
+
     private(set) var currentStatus: SiteAssemblyStatus = .idle {
         didSet {
             DispatchQueue.main.async { [weak self] in
-                guard let strongSelf = self else {
+                guard let self = self else {
                     return
                 }
-                strongSelf.statusChangeHandler?(strongSelf.currentStatus)
+                self.statusChangeHandler?(self.currentStatus)
             }
         }
     }
 
-    private(set) var statusChangeHandler: SiteAssemblyStatusChangedHandler?
+    /// This method serves as the entry point for creating an enhanced site. It consists of a few steps:
+    /// 1. The creation request is first validated.
+    /// 1. If a site passes validation, a new service invocation triggers creation.
+    /// 1. The details of the created Blog are persisted to the client.
+    ///
+    /// Previously the site's theme & tagline were set post-creation. This is now handled on the server via Headstart.
+    ///
+    /// - Parameters:
+    ///   - creationRequest:    the request with which to initiate site creation.
+    ///   - changeHandler:      this closure is invoked when site assembly status changes occur.
+    ///
+    func createSite(creationRequest: SiteCreationRequest, changeHandler: SiteAssemblyStatusChangedHandler? = nil) {
+        self.creationRequest = creationRequest
+        self.statusChangeHandler = changeHandler
 
-    // MARK: SiteAssemblyService
+        beginAssembly()
+        validatePendingRequest()
+    }
 
-    func createSite(creatorOutput assemblyInput: SiteCreatorOutput, changeHandler: SiteAssemblyStatusChangedHandler? = nil) {
-        self.statusChangeHandler = changeHandler
+    // MARK: Private behavior
+
+    private func beginAssembly() {
+        currentStatus = .inProgress
+    }
+
+    private func endFailedAssembly() {
```

Comment:
Curious about the naming here. Why "end" and not just failedAssembly?

Reply:
I named this method and `endSuccessfulAssembly` below (line 89) to provide a modicum of symmetry with `beginAssembly` (line 81) above. 

Of course, since this method is marked `private`, we can name it anything we'd like. 😁 

I'm willing to change this, but I'd like to make sure that each of these method names changes in concert. I welcome suggestions here.

Reply:
Interesting. I don't feel strongly about this so please name as you like but I think the successful and failed implies the end of the assembly so the end just made me wonder but totally up to you